app/src/main/java/io/appium/settings/RecorderThread.java

```diff
@@ -0,0 +1,419 @@
+                .setSampleRate(sampleRate)
+                .setChannelMask(AudioFormat.CHANNEL_IN_MONO)
+                .build();
+
+        AudioRecord.Builder audioRecordBuilder = new AudioRecord.Builder();
+        AudioPlaybackCaptureConfiguration apcc =
+                new AudioPlaybackCaptureConfiguration.Builder(this.mediaProjection)
+                        .addMatchingUsage(AudioAttributes.USAGE_MEDIA)
+                        .build();
+        audioRecord = audioRecordBuilder.setAudioFormat(newAudioFormat)
+                .setBufferSizeInBytes(4 * minBufferSize)
+                .setAudioPlaybackCaptureConfig(apcc)
+                .build();
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void startAudioRecord() {
+        audioRecordThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                Thread.currentThread().setPriority(Thread.MAX_PRIORITY);
+                try {
+                    audioRecord.startRecording();
+                } catch (Exception e) {
+                    asyncError = true;
+                    e.printStackTrace();
+                    return;
+                }
+                try {
+                    while (!audioStopped) {
+                        int index = audioEncoder.dequeueInputBuffer(
+                                RecorderConstant.MEDIA_QUEUE_BUFFERING_DEFAULT_TIMEOUT);
+                        if (index < 0) {
+                            continue;
+                        }
+                        ByteBuffer inputBuffer = audioEncoder.getInputBuffer(index);
+                        if (inputBuffer == null) {
+                            if (!stopped) {
+                                asyncError = true;
+                            }
+                            return;
+                        }
+                        inputBuffer.clear();
+                        int read = audioRecord.read(inputBuffer, inputBuffer.capacity());
+                        if (read < 0) {
+                            if (!stopped) {
+                                asyncError = true;
+                            }
+                            break;
+                        }
+                        audioEncoder.queueInputBuffer(index, 0, read,
+                                getPresentationTimeUs(), 0);
+                    }
+                } catch (Exception e) {
+                    if (!stopped) {
+                        Log.e(TAG, "Audio error", e);
+                        asyncError = true;
+                        e.printStackTrace();
+                    }
+                } finally {
+                    audioRecord.stop();
+                    audioRecord.release();
+                    audioRecord = null;
+                }
+            }
+        });
+        audioRecordThread.start();
+    }
+
+    private long getPresentationTimeUs() {
```

Comment:
this could be simplified by making startTimestampInitialized of type Optional or at least Long. So we can clearly see when it is not initialized yet (e.g. equals Optional.empty() or null)

Reply:
I couldn't fully understand this one, can you give a detail a bit? thanks 👍

Reply:
currently `-1` is used as so-called "guard value". This might be not very obvious as there is no clear indication about why it was selected being magic number.  
To make this more obvious one would make use of Optional. Even making the guard value to a constant would already significantly improve the readability. For example, compare the following:

```
long x = -1;
long x = NO_TIMESTAMP_SET;
Optional<Long> x = Optional.empty(); // then y = x.getOrElseThrow(TimestampNotInitializedError::new)
```

Reply:
hmm got it but `startTimestampInitialized` is currently boolean value, i couldn't see the  `-1` value anywhere in this `getPresentationTimeUs` method, i now changed name to `isStartTimestampInitialized` and initialized to false as default, renaming helps a bit for readability but i couldn't understand the win by making current `boolean` type to `Optional<boolean>` (we already init as false so no null check or exception-throwing check is required) sorry for misunderstanding

Reply:
okay done